presto-clp/src/main/java/com/facebook/presto/plugin/clp/ClpPlanOptimizer.java

```diff
@@ -0,0 +1,115 @@
+import com.facebook.presto.spi.plan.PlanNode;
+import com.facebook.presto.spi.plan.PlanNodeIdAllocator;
+import com.facebook.presto.spi.plan.TableScanNode;
+import com.facebook.presto.spi.relation.RowExpression;
+import com.facebook.presto.spi.relation.VariableReferenceExpression;
+
+import java.util.Map;
+import java.util.Optional;
+
+import static com.facebook.presto.spi.ConnectorPlanRewriter.rewriteWith;
+
+public class ClpPlanOptimizer
+        implements ConnectorPlanOptimizer
+{
+    private static final Logger log = Logger.get(ClpPlanOptimizer.class);
+    private final FunctionMetadataManager functionManager;
+    private final StandardFunctionResolution functionResolution;
+
+    public ClpPlanOptimizer(FunctionMetadataManager functionManager,
+                            StandardFunctionResolution functionResolution)
+    {
+        this.functionManager = functionManager;
+        this.functionResolution = functionResolution;
+    }
+
+    @Override
+    public PlanNode optimize(PlanNode maxSubplan,
+                             ConnectorSession session,
+                             VariableAllocator variableAllocator,
+                             PlanNodeIdAllocator idAllocator)
+    {
+        return rewriteWith(new Rewriter(idAllocator), maxSubplan);
+    }
+
+    private class Rewriter
+            extends ConnectorPlanRewriter<Void>
+    {
+        private final PlanNodeIdAllocator idAllocator;
+
+        public Rewriter(PlanNodeIdAllocator idAllocator)
+        {
+            this.idAllocator = idAllocator;
+        }
+
+        @Override
+        public PlanNode visitFilter(FilterNode node, RewriteContext<Void> context)
+        {
+            if (!(node.getSource() instanceof TableScanNode)) {
+                return node;
+            }
+
+            TableScanNode tableScanNode = (TableScanNode) node.getSource();
+            Map<VariableReferenceExpression, ColumnHandle> assignments = tableScanNode.getAssignments();
+            TableHandle tableHandle = tableScanNode.getTable();
+            ClpTableHandle clpTableHandle = (ClpTableHandle) tableHandle.getConnectorHandle();
+            ClpExpression clpExpression = node.getPredicate()
+                    .accept(new ClpFilterToKqlConverter(functionResolution, functionManager, assignments),
+                            null);
+            Optional<String> kqlQuery = clpExpression.getDefinition();
+            Optional<RowExpression> remainingPredicate = clpExpression.getRemainingExpression();
+            if (!kqlQuery.isPresent()) {
+                return node;
+            }
+            log.debug("KQL query: %s", kqlQuery.get());
+            ClpTableLayoutHandle clpTableLayoutHandle = new ClpTableLayoutHandle(clpTableHandle, kqlQuery);
+            TableScanNode newTableScanNode = new TableScanNode(
+                    tableScanNode.getSourceLocation(),
+                    idAllocator.getNextId(),
+                    new TableHandle(
```

Comment:
why do we need to create a new table handle here? can we just used the tableHandle?  is it just for the kqlQuery?

Reply:
Yeah, that's because we create a new `ClpTableLayoutHandle` with the KQL query. Similar to https://github.com/prestodb/presto/blob/master/presto-base-jdbc/src/main/java/com/facebook/presto/plugin/jdbc/optimization/JdbcComputePushdown.java#L128 and https://github.com/prestodb/presto/blob/c87529695ff0ff2375e52d46c9a7149137831537/presto-druid/src/main/java/com/facebook/presto/druid/DruidPlanOptimizer.java#L162